lib/TPP/Dialect/Xsmm/XsmmUtils.cpp

```diff
@@ -287,6 +358,175 @@ FailureOr<BinaryFlags> getBinaryFlags(Type operandType, Type outputType,
   return failure();
 }
 
+FailureOr<int64_t> getLeadingDim(Type type, size_t pos) {
+  // Not shaped type, the leading dimension is the single scalar.
+  auto memref = dyn_cast<MemRefType>(type);
+  if (!memref)
+    return 1;
+  // For 1d memref we cannot use the stride as leading dimension, but the
+  // leading dimension is the dimension itself.
+  if (memref.getRank() == 1)
+    return memref.getShape()[0];
+
+  SmallVector<int64_t> strides;
+  int64_t offset;
+  if (failed(getStridesAndOffset(memref, strides, offset)))
+    return failure();
+  // fail if the strides are non-constant
+  if (llvm::any_of(strides, [](int64_t stride) {
+        return stride == ShapedType::kDynamic;
+      }))
+    return failure();
+  return strides[pos];
+}
+
+bool isTwoDTransposeOp(vector::TransposeOp transposeOp) {
+  auto operandType = dyn_cast<VectorType>(transposeOp.getOperand().getType());
+  bool isVnni = vnni::utils::isInVnniLayout(operandType.getRank(), operandType);
+  if (isVnni || operandType.getRank() != 2 ||
+      dyn_cast<VectorType>(transposeOp.getResult().getType()).getRank() != 2 ||
+      (isa<scf::ForallOp>(transposeOp->getParentOp()) &&
+       dyn_cast<scf::ForallOp>(transposeOp->getParentOp()).getRank() != 2))
+    return false;
+  return true;
+}
+
+// Extract the operands to be used in the function call. For each memref operand
+// extract the aligned pointer and the offset.
+SmallVector<Value> getOperands(OpBuilder &builder, Location loc,
+                               ValueRange operands, IntegerAttr dataTypeAttr,
+                               Operation *parentOp, bool getResults) {
+  SmallVector<Value> res;
+  builder.setInsertionPoint(parentOp);
+  for (Value operand : operands) {
+    auto memrefType = dyn_cast<MemRefType>(operand.getType());
+    if (!memrefType) {
+      res.push_back(operand);
+      continue;
+    }
+    auto [ptr, offset] = ::mlir::utils::getPtrAndOffset(builder, operand, loc);
+    res.push_back(ptr);
+    res.push_back(offset);
+  }
+  return res;
+}
+
+SmallVector<Type> extractInvokeOperandTypes(OpBuilder &builder,
+                                            ValueRange operands) {
+  SmallVector<Type> results;
+  // One extra operand for datatype
+  for (Value operand : operands) {
+    Type operandType = operand.getType();
+    if (auto memrefType = dyn_cast<MemRefType>(operandType)) {
+      // TODO: non-POD will require an LLVMTypeConverter.
+      Type basePtrType = LLVM::LLVMPointerType::get(builder.getContext());
+      results.push_back(basePtrType);
+      results.push_back(builder.getIndexType()); // offset
+    } else {
+      results.push_back(operand.getType());
+    }
+  }
+  return results;
+}
+
+int64_t getOredFlags(ArrayAttr flags) {
```

Comment:
This sounds like a GEMM function, but you use for the unary check below. I'm not sure what's going on.

Reply:
I realize that, the flag is always unary::none so we don't need oredflags function. I'll get rid of it.

Reply:
Just as a side note, getOredFlags is used on unaryFlags in ConvertXsmmToFunc which is why I ended up using the same utility. We'll need to update that as well, till such a time that we retire that conversion.